src/pydata_sphinx_theme/assets/scripts/index.js

```diff
@@ -71,7 +79,74 @@ function scrollToActive() {
   });
 }
 
+////////////////////////////////////////////////////////////////////////////////
+// Theme interaction
+////////////////////////////////////////////////////////////////////////////////
+var prefersDark = window.matchMedia("(prefers-color-scheme: dark)");
+
+function autoTheme(e) {
+  document.body.dataset.theme = prefersDark.matches ? "dark" : "light";
+}
+
+function setTheme(mode) {
+  if (mode !== "light" && mode !== "dark" && mode !== "auto") {
+    console.error(`Got invalid theme mode: ${mode}. Resetting to auto.`);
+    mode = "auto";
+  }
+
+  // get the theme
+  var colorScheme = prefersDark.matches ? "dark" : "light";
+  document.body.dataset.mode = mode;
+  document.body.dataset.theme = mode == "auto" ? colorScheme : mode;
+
+  // save mode
+  localStorage.setItem("theme", mode);
+  console.log(`Changed to ${mode} mode.`);
+
+  // add a listener if set on auto
+  prefersDark.onchange = mode == "auto" ? autoTheme : "";
+}
+
+function cycleTheme() {
+  const defaultMode = document.body.dataset.defaultMode || "auto";
+  const currentTheme = localStorage.getItem("theme") || defaultMode;
+
+  if (prefersDark.matches) {
+    // Auto (dark) -> Light -> Dark
+    if (currentTheme === "auto") {
+      setTheme("light");
+    } else if (currentTheme == "light") {
+      setTheme("dark");
+    } else {
+      setTheme("auto");
+    }
+  } else {
+    // Auto (light) -> Dark -> Light
+    if (currentTheme === "auto") {
+      setTheme("dark");
+    } else if (currentTheme == "dark") {
+      setTheme("light");
+    } else {
+      setTheme("auto");
+    }
+  }
+}
+
+function setupTheme() {
+  // setup at least one time
+  const defaultMode = document.body.dataset.defaultMode || "auto";
+  const currentTheme = localStorage.getItem("theme") || defaultMode;
+  setTheme(currentTheme);
```

Comment:
Is there a way that we can embed the default theme without relying on a JavaScript call? When we do it in JS like here, it creates a flicker every time somebody switches pages:

![chrome_ZxO9DOwkaz](https://user-images.githubusercontent.com/1839645/153071592-e5ef5d7e-7fca-4dbb-a246-7e4a1ae8e178.gif)

Maybe the Furo theme has inspiration for how to avoid this?

Reply:
The furo theme is flickering as well, the background is just preshot (look very intensly the theme switcher and you'll see it). Looking on the web it seems that the best solution is to execute the theme selection script in the header before the rendering of the page. Then the theme will be applied on the `<html>` tag instead of `<body>`.

Reply:
@choldgraf If i moved the control to the `html` tag i' will also loose the "default_mode" feature as I cannot add extra parameter in  the `html` tag from the layout.html file. It will default to "auto" as in the Furo theme. I don't see any problem I just want to make sur eyou agrees

Reply:
I think it's fine to default to `auto` - is that how most other dark modes behave? As long as we can remain within user expectations I think it's fine!

Reply:
> The furo theme is flickering as well, the background is just preshot

Wait, what? That shouldn't be the case -- the render should be blocked until the JS adds the class that the relevant CSS stylises.

Reply:
When wrote it I was doing the check with my poor eyes...
> (look very intensly the theme switcher and you'll see it)

Now I'm a better person and I discover that browsers have a chronological analyser. I just load the furo theme documentation and the JS is loaded prior to anything so no flickering on your side: 
<img width="807" alt="Capture d’écran 2022-02-16 à 23 51 02" src="https://user-images.githubusercontent.com/12596392/154371000-57d2b8d8-c879-4aa4-8793-a2ef3a76c26f.png">

Reply:
This can be marked as resolved.